File: packages/modules/ImsMedia/tests/native/service/src/com/android/telephony/imsmedia/lib/libimsmedia/config/EvsParamsTest.cpp
/**
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <EvsParams.h>
#include <gtest/gtest.h>

using namespace android::telephony::imsmedia;

const int32_t kEvsBandwidth = EvsParams::EVS_BAND_NONE;
const int32_t kEvsMode = 8;
const int8_t kChannelAwareMode = 3;
const bool kUseHeaderFullOnly = false;
const bool kcodecModeRequest = 15;

TEST(EvsParamsTest, TestGetterSetter)
{
    EvsParams* param = new EvsParams();
    param->setEvsBandwidth(kEvsBandwidth);
    param->setEvsMode(kEvsMode);
    param->setChannelAwareMode(kChannelAwareMode);
    param->setUseHeaderFullOnly(kUseHeaderFullOnly);
    param->setCodecModeRequest(kcodecModeRequest);
    EXPECT_EQ(param->getEvsMode(), kEvsMode);
    EXPECT_EQ(param->getChannelAwareMode(), kChannelAwareMode);
    EXPECT_EQ(param->getUseHeaderFullOnly(), kUseHeaderFullOnly);
    EXPECT_EQ(param->getCodecModeRequest(), kcodecModeRequest);
    delete param;
}

TEST(EvsParamsTest, TestParcel)
{
    EvsParams* param = new EvsParams();
    param->setEvsBandwidth(kEvsBandwidth);
    param->setEvsMode(kEvsMode);
    param->setChannelAwareMode(kChannelAwareMode);
    param->setUseHeaderFullOnly(kUseHeaderFullOnly);
    param->setCodecModeRequest(kcodecModeRequest);

    android::Parcel parcel;
    param->writeToParcel(&parcel);
    parcel.setDataPosition(0);

    EvsParams* param2 = new EvsParams();
    param2->readFromParcel(&parcel);
    EXPECT_EQ(*param2, *param);

    delete param;
    delete param2;
}

TEST(EvsParamsTest, TestAssign)
{
    EvsParams param;
    param.setEvsBandwidth(kEvsBandwidth);
    param.setEvsMode(kEvsMode);
    param.setChannelAwareMode(kChannelAwareMode);
    param.setUseHeaderFullOnly(kUseHeaderFullOnly);
    param.setCodecModeRequest(kcodecModeRequest);

    EvsParams param2;
    param2 = param;
    EXPECT_EQ(param, param2);
}

TEST(EvsParamsTest, TestEqual)
{
    EvsParams* param = new EvsParams();
    param->setEvsBandwidth(kEvsBandwidth);
    param->setEvsMode(kEvsMode);
    param->setChannelAwareMode(kChannelAwareMode);
    param->setUseHeaderFullOnly(kUseHeaderFullOnly);
    param->setCodecModeRequest(kcodecModeRequest);

    EvsParams* param2 = new EvsParams();
    param2->setEvsBandwidth(kEvsBandwidth);
    param2->setEvsMode(kEvsMode);
    param2->setChannelAwareMode(kChannelAwareMode);
    param2->setUseHeaderFullOnly(kUseHeaderFullOnly);
    param2->setCodecModeRequest(kcodecModeRequest);

    EXPECT_EQ(*param, *param2);
    delete param;
    delete param2;
}

TEST(EvsParamsTest, TestNotEqual)
{
    EvsParams* param = new EvsParams();
    param->setEvsBandwidth(kEvsBandwidth);
    param->setEvsMode(kEvsMode);
    param->setChannelAwareMode(kChannelAwareMode);
    param->setUseHeaderFullOnly(kUseHeaderFullOnly);
    param->setCodecModeRequest(kcodecModeRequest);

    EvsParams* param2 = new EvsParams();
    param2->setEvsBandwidth(kEvsBandwidth);
    param2->setEvsMode(5);
    param2->setChannelAwareMode(kChannelAwareMode);
    param2->setUseHeaderFullOnly(kUseHeaderFullOnly);
    param2->setCodecModeRequest(kcodecModeRequest);

    EvsParams* param3 = new EvsParams();
    param3->setEvsBandwidth(kEvsBandwidth);
    param3->setEvsMode(kEvsMode);
    param3->setChannelAwareMode(kChannelAwareMode);
    param3->setUseHeaderFullOnly(true);
    param3->setCodecModeRequest(kcodecModeRequest);

    EvsParams* param4 = new EvsParams();
    param4->setEvsBandwidth(kEvsBandwidth);
    param4->setEvsMode(kEvsMode);
    param4->setChannelAwareMode(kChannelAwareMode);
    param4->setUseHeaderFullOnly(kUseHeaderFullOnly);
    param4->setCodecModeRequest(14);

    EXPECT_NE(*param, *param2);
    EXPECT_NE(*param, *param3);
    EXPECT_NE(*param, *param4);

    delete param;
    delete param2;
    delete param3;
    delete param4;
}

File: packages/modules/ImsMedia/tests/unit/src/com/android/telephony/imsmedia/EvsParamsTest.java
/**
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.telephony.imsmedia.tests;

import static com.google.common.truth.Truth.assertThat;

import android.os.Parcel;
import android.telephony.imsmedia.EvsParams;

import androidx.test.runner.AndroidJUnit4;

import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
public class EvsParamsTest {
    private static final byte CHANNEL_AWARE_MODE = 7;
    private static final byte CODEC_MODE_REQUEST = 15;
    private static final boolean USE_HEADER_FULL_ONLY = true;

    private EvsParams createEvsParams(
            final int evsBandwidth,
            final int evsMode,
            final byte channelAwareMode,
            final boolean useHeaderFullOnly,
            final byte codecModeRequest) {
        return new EvsParams.Builder()
                .setEvsbandwidth(evsBandwidth)
                .setEvsMode(evsMode)
                .setChannelAwareMode(channelAwareMode)
                .setHeaderFullOnly(useHeaderFullOnly)
                .setCodecModeRequest(codecModeRequest)
                .build();
    }

    @Test
    public void testConstructorAndGetters() {
        EvsParams evs = createEvsParams(EvsParams.EVS_WIDE_BAND, EvsParams.EVS_MODE_7,
                CHANNEL_AWARE_MODE, USE_HEADER_FULL_ONLY, CODEC_MODE_REQUEST);

        assertThat(evs.getEvsBandwidth()).isEqualTo(EvsParams.EVS_WIDE_BAND);
        assertThat(evs.getEvsMode()).isEqualTo(EvsParams.EVS_MODE_7);
        assertThat(evs.getChannelAwareMode()).isEqualTo(CHANNEL_AWARE_MODE);
        assertThat(evs.getUseHeaderFullOnly()).isEqualTo(USE_HEADER_FULL_ONLY);
        assertThat(evs.getCodecModeRequest()).isEqualTo(CODEC_MODE_REQUEST);
    }

    @Test
    public void testParcel() {
        EvsParams evs = createEvsParams(EvsParams.EVS_WIDE_BAND, EvsParams.EVS_MODE_7,
                CHANNEL_AWARE_MODE, USE_HEADER_FULL_ONLY, CODEC_MODE_REQUEST);

        Parcel parcel = Parcel.obtain();
        evs.writeToParcel(parcel, 0);
        parcel.setDataPosition(0);

        EvsParams parcelConfig = EvsParams.CREATOR.createFromParcel(parcel);
        assertThat(evs).isEqualTo(parcelConfig);
    }

    @Test
    public void testEqual() {
        EvsParams evs1 = createEvsParams(EvsParams.EVS_WIDE_BAND, EvsParams.EVS_MODE_7,
                CHANNEL_AWARE_MODE, USE_HEADER_FULL_ONLY, CODEC_MODE_REQUEST);

        EvsParams evs2 = createEvsParams(EvsParams.EVS_WIDE_BAND, EvsParams.EVS_MODE_7,
                CHANNEL_AWARE_MODE, USE_HEADER_FULL_ONLY, CODEC_MODE_REQUEST);

        assertThat(evs1).isEqualTo(evs2);
    }

    @Test
    public void testNotEqual() {
        EvsParams evs1 = createEvsParams(EvsParams.EVS_WIDE_BAND, EvsParams.EVS_MODE_7,
                CHANNEL_AWARE_MODE, USE_HEADER_FULL_ONLY, CODEC_MODE_REQUEST);

        EvsParams evs2 = createEvsParams(EvsParams.EVS_WIDE_BAND, EvsParams.EVS_MODE_6,
                CHANNEL_AWARE_MODE, USE_HEADER_FULL_ONLY, CODEC_MODE_REQUEST);

        assertThat(evs1).isNotEqualTo(evs2);

        EvsParams evs3 = createEvsParams(EvsParams.EVS_WIDE_BAND, EvsParams.EVS_MODE_7,
                (byte) 8, USE_HEADER_FULL_ONLY, CODEC_MODE_REQUEST);

        assertThat(evs1).isNotEqualTo(evs3);

        EvsParams evs4 = createEvsParams(EvsParams.EVS_WIDE_BAND, EvsParams.EVS_MODE_7,
                CHANNEL_AWARE_MODE, false, CODEC_MODE_REQUEST);

        assertThat(evs1).isNotEqualTo(evs4);

        EvsParams evs6 = createEvsParams(EvsParams.EVS_SUPER_WIDE_BAND, EvsParams.EVS_MODE_7,
                CHANNEL_AWARE_MODE, USE_HEADER_FULL_ONLY, CODEC_MODE_REQUEST);

        assertThat(evs1).isNotEqualTo(evs6);

        EvsParams evs7 = createEvsParams(EvsParams.EVS_WIDE_BAND, EvsParams.EVS_MODE_7,
                CHANNEL_AWARE_MODE, USE_HEADER_FULL_ONLY, (byte) 14);

        assertThat(evs1).isNotEqualTo(evs7);
    }
}
